### PR TITLE
Readability enhancements for line wrapping

### PR DESCRIPTION
**Pull request by whereswaldon**
This PR is against the `iterator` branch (the base of #68) instead of `main`. The reason I've separated it out is that I didn't want to put an overwhelming quantity of change into #68, but this work does depend on the changes in that branch.

These changes are motivated by wanting to introduce wrapping on grapheme cluster boundaries as well as "word" boundaries (what we do right now). I needed to reorganize our logic some in order to achieve that, so I was simply trying to refactor so that I could reuse appropriate logic and make things more readable. I did _not_ expect a significant speedup from this refactor, but...

```
                                      │ with-scratch.txt │        after-refactor.txt         │
                                      │      sec/op      │   sec/op     vs base              │
Wrapping/10runes-arabic-1parts-8             31.39n ± 4%   28.78n ± 2%  -8.33% (p=0.002 n=6)
Wrapping/10runes-arabic-10parts-8            2.748µ ± 4%   2.551µ ± 0%  -7.20% (p=0.002 n=6)
Wrapping/100runes-arabic-1parts-8            18.66µ ± 3%   17.52µ ± 1%  -6.07% (p=0.002 n=6)
Wrapping/100runes-arabic-10parts-8           19.66µ ± 4%   18.22µ ± 1%  -7.30% (p=0.002 n=6)
Wrapping/100runes-arabic-100parts-8          36.90µ ± 3%   34.80µ ± 1%  -5.69% (p=0.002 n=6)
Wrapping/1000runes-arabic-1parts-8           185.1µ ± 4%   167.4µ ± 0%  -9.55% (p=0.002 n=6)
Wrapping/1000runes-arabic-10parts-8          184.0µ ± 3%   169.1µ ± 0%  -8.13% (p=0.002 n=6)
Wrapping/1000runes-arabic-100parts-8         191.4µ ± 3%   179.9µ ± 0%  -5.99% (p=0.002 n=6)
Wrapping/1000runes-arabic-1000parts-8        370.9µ ± 7%   356.4µ ± 1%  -3.92% (p=0.002 n=6)
Wrapping/10runes-latin-1parts-8              31.00n ± 2%   28.52n ± 3%  -7.98% (p=0.002 n=6)
Wrapping/10runes-latin-10parts-8             2.362µ ± 5%   2.139µ ± 0%  -9.46% (p=0.002 n=6)
Wrapping/100runes-latin-1parts-8             14.83µ ± 3%   13.90µ ± 0%  -6.28% (p=0.002 n=6)
Wrapping/100runes-latin-10parts-8            16.11µ ± 4%   14.94µ ± 1%  -7.27% (p=0.002 n=6)
Wrapping/100runes-latin-100parts-8           33.58µ ± 5%   31.66µ ± 2%  -5.73% (p=0.002 n=6)
Wrapping/1000runes-latin-1parts-8            149.2µ ± 5%   136.8µ ± 3%  -8.30% (p=0.002 n=6)
Wrapping/1000runes-latin-10parts-8           156.2µ ± 5%   141.9µ ± 1%  -9.17% (p=0.002 n=6)
Wrapping/1000runes-latin-100parts-8          169.1µ ± 3%   156.0µ ± 1%  -7.76% (p=0.002 n=6)
Wrapping/1000runes-latin-1000parts-8         357.7µ ± 4%   336.6µ ± 0%  -5.89% (p=0.002 n=6)
WrappingHappyPath-8                          30.37n ± 4%   28.02n ± 1%  -7.75% (p=0.002 n=6)
geomean                                      15.95µ        14.79µ       -7.26%
```

There are no changes at all to memory use or allocations.

Anyway, since this work depends on #68, and is logically somewhat distinct, I've opened it as a separate PR. I look forward to your review on both. :D

### CODE REVIEW

File: shaping/wrapping.go [hunk 1]
```diff
@@ -557,36 +557,77 @@ const (
...
+}
+
+// lineBuilder manages line wrapping candidates.
+type lineBuilder struct {
+	// best tracks the best known line built by this builder within the maxWidth it was used with.
+	best []Output
+	// candidate holds intermediate line state as new lines are built and tested to fit.
+	candidate []Output
+	// candidateWidth tracks the total advance of the runs in candidate.
+	candidateWidth fixed.Int26_6
```
Comment (benoitkugler) [inline shaping/wrapping.go:578]: Maybe add something like 
it is the cached value of sum(output.Advance for output in candidate)

File: shaping/wrapping.go [hunk 2]
```diff
@@ -557,36 +557,77 @@ const (
...
+	l.candidate = append(l.candidate, run)
+	l.candidateWidth += run.Advance
+}
+
+// commitCandidate updates l.best to be the results of concatenating withNewRun to
+// the end of l.candidate, but does not modify l.candidate. This allows us to save
+// a known good line while also perserving our ability to try other suffixes on the
+// end of l.candidate.
+func (l *lineBuilder) commitCandidate(withNewRun Output) {
+	l.best = commitCandidate(l.best, l.candidate, withNewRun)
```
Comment (benoitkugler) [inline shaping/wrapping.go:601]: Maybe merge `commitCandidate` into this method ? Unless you expect to use it later ?
I also note that `commitCandidate` is only used with a single `Output` (again, for now) so you could simplify its signature and implementation.

File: shaping/wrapping.go [hunk 3]
```diff
@@ -645,82 +688,114 @@ func (l *LineWrapper) WrapNextLine(maxWidth int, scratch LineScratch) (finalLine
...
+	endLine
+	// newLine indicates that the returned best []Output contains a full line that does not terminate the text.
+	newLine
+	// fits indicates that the text up to the break option fit within the line and that another break
+	// option can be attempted.
+	fits
+	// cannotFit indicates that this is the first break option on the line, and that even this option cannot
+	// fit within the space available.
+	cannotFit
+)
```
Comment (benoitkugler) [inline shaping/wrapping.go:740]: These constants add so much clarity ! Very nice improvement :)

**Review by benoitkugler — CHANGES_REQUESTED** [review]
Nicely done !

I've left some minor suggestions.

File: shaping/wrapping.go (hunk 2 above)
Comment (whereswaldon) [reply shaping/wrapping.go:601]: True, I think it would probably make sense to move that here. Thanks!

**Comment by whereswaldon** [comment]
I've changed my mind about splitting this changeset from #68, so I've incorporated these changes there.

**Pull request closed without merging**